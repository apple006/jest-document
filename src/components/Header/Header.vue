<template>
  <header class="header">
    <div class="header__img">
      <img src="@/assets/logo.jpg" alt="" height="100%">
    </div>
    <div class="header__input">
      <input type="text" class="search" data-test="input" v-model="value">
      <button class="button__add" data-test="button" @click="handleAddBtn">添 加</button>
    </div>
  </header>
</template>

<script>
import { mapState, mapMutations } from 'vuex'

export default {
  data() {
    return {
      value: ''
    }
  },
  methods: {
    ...mapMutations(['addListItem']),
    handleAddBtn() {
      if (this.value !== '') {
        this.addListItem(this.value)
        this.value = ''
      }
    }
  }
}
</script>

<style lang="sass" scoped>
.header
  margin: 0 auto;
  display: flex;
  justify-content: center;
  width: 800px;

.header__img
  height: 80px

.header__input
  margin-left: 30px;
  display: flex;
  width: 470px;
  align-items: center;

.search
  outline: none;
  width: 382px;
  height: 42px;
  border: 2px solid #00d0d4;
  border-radius: 5px 0 0 5px;
  display: inline-block;
  padding-left: 16px;

.button__add
  border: none;
  outline: none;
  width: 68px;
  height: 48px;
  font-size: 20px;
  color: #fff;
  cursor: pointer;
  background: linear-gradient(116deg,#00d0d4 5%,#00c5c3 51%,#9effa4 100%);
</style>
