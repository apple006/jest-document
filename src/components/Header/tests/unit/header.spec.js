import { shallowMount, createLocalVue } from '@vue/test-utils'
import Header from '@/components/Header/Header'
import Vuex from 'vuex'

const localVue = createLocalVue()
localVue.use(Vuex)
let store, mutations

beforeEach(() => {
  // 每次运行测试实例的时候重新生成一个store，避免测试实例之间相互影响
  mutations = {
    addListItem: jest.fn(),
    minusListItem: jest.fn()
  }
  store = new Vuex.Store({
    state: {
      list: []
    },
    mutations
  })
})

describe('Header.vue', () => {
  it('组件含有输入框', () => {
    const wrapper = shallowMount(Header)
    const input = wrapper.find('[data-test="input"]')
    expect(input.exists()).toBeTruthy()
  })

  it('输入框初始值为空', () => {
    const wrapper = shallowMount(Header)
    const value = wrapper.vm.value
    expect(value).toBe('')
  })

  it('当输入框有内容时，点击 添加 按钮时，数据改变，并清空输入框', () => {
    const wrapper = shallowMount(Header, { store, localVue })
    const input = wrapper.find('[data-test="input"]')
    const button = wrapper.find('[data-test="button"]')

    input.setValue('添加一条数据')
    button.trigger('click')
    expect(mutations.addListItem).toHaveBeenCalled()
    expect(wrapper.vm.value).toBe('')
  })

  it('当输入框没有内容时，点击 添加 按钮时，没反应', () => {
    const wrapper = shallowMount(Header, { store, localVue })
    const input = wrapper.find('[data-test="input"')
    const button = wrapper.find('[data-test="button"]')

    input.setValue('')
    button.trigger('click')
    expect(mutations.addListItem).not.toHaveBeenCalled()
  })

  it('组件发生改变', () => {
    const wrapper = shallowMount(Header)
    expect(wrapper).toMatchSnapshot()
  })

})