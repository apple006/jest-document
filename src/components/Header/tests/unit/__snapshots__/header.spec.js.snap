// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header.vue 组件发生改变 1`] = `
<header class="header">
  <div class="header__img"><img src="@/assets/logo.jpg" alt="" height="100%"></div>
  <div class="header__input"><input type="text" data-test="input" class="search"> <button data-test="button" class="button__add">添 加</button></div>
</header>
`;
